components/in_memory_engine/src/region_stats.rs

```diff
@@ -269,68 +215,96 @@ impl RegionStatsManager {
-                        } else {
-                            Some((r, s))
-                        }
-                    })
-                    .rev() // evict the regions with least next + prev after the filter
-                    .take(max_count_to_evict)
-                    .collect();
+        if cached_region_stats.len() <= MIN_REGION_COUNT_TO_EVICT {
+            return (regions_to_load, vec![]);
+        }
 
-            // TODO(SpadeA): remove it after it's stable
-            let debug: Vec<_> = region_to_evict
-                .iter()
-                .map(|(r, s)| {
-                    format!(
-                        "region_id={}, cop={}, cop_detail={:?}, mvcc_amplification={}",
-                        r.id,
-                        s.query_stats.coprocessor,
-                        s.cop_detail,
-                        s.cop_detail.mvcc_amplification(),
-                    )
-                })
-                .collect();
-            info!(
-                "ime collect regions to evict";
-                "reached_stop_limit" => reach_stop_load,
-                "regions" => ?debug,
-            );
-            region_to_evict.into_iter().map(|(r, _)| r).collect()
-        } else {
-            vec![]
-        };
+        let mut total_next_prev = 0;
+        let mut total_cop_requests = 0;
+        for crs in &cached_region_stats {
+            total_next_prev += crs.stat.cop_detail.iterated_count();
+            total_cop_requests += crs.sma_cop_requests_avg;
+        }
+        let avg_next_prev = total_next_prev / cached_region_stats.len();
+        let avg_cop_requests = total_cop_requests / cached_region_stats.len();
+        let is_memory_low = memory_controller.reached_stop_load_threshold();
+
+        let now = Instant::now();
+        let mut region_to_evict = Vec::with_capacity(max_count_to_evict);
+        for crs in cached_region_stats.into_iter().take(max_count_to_evict) {
+            let need_evict = {
+                // In case, memory usage is relatively low, we evict those that
+                // should not be cached.
+                //
+                // NB: When a region is cached, its RegionWriteCfCopDetail only
+                // reflects the MVCC amplification in IME not in RocksDB. So
+                // low MVCC amplification is not a good indicator for eviction.
+                //
+                // Instead, we assume workload patterns remain consistent after
+                // caching and evict based on request frequency.
+                let is_cop_requests_low = crs.sma_cop_requests_avg
+                    <= avg_cop_requests / SMA_COP_REQUEST_AVG_FILTER_FACTOR;
+
+                let is_cop_requests_reliable =
+                    crs.sma_cop_requests_count > SMA_COP_REQUEST_COUNT_FILTER;
+
+                let is_iterated_count_low = crs.stat.cop_detail.iterated_count()
+                    <= avg_next_prev / ITERATED_COUNT_FILTER_FACTOR;
+
+                is_memory_low
```

Comment:
Does it should be `!is_memory_low` to represent that although the memory_usage is low, it should also evict some cached_regions which are not frequently accessed ?

Reply:
The variable name is misleading, it actually means the remain memory capacity for IME is low. Will revise the name.